### DISCUSSION TITLE
failed to build: no such subcommand: `wasi`

### DISCUSSION THREAD

**Discussion by ekeren** [winglang/wing#522] [Q&A] [answered]
When running  `npm run build` (after I ran `npm install`) I get the following error message:

```sh
eyalkeren@eyals-MacBook-Pro wing % npm run build             

> build
> nx run-many --target=build --all=true


    ✔  nx run wing-language-server:build-native  [existing outputs match the cache, left as is]
    ✔  nx run wing-api-checker:build  [existing outputs match the cache, left as is]

    ✖  nx run wingc:build-wasm
       error: no such subcommand: `wasi`
       
        Did you mean `add`?
       
        View all installed commands with `cargo --list`
       
        >  NX   ERROR: Something went wrong in run-commands - Command failed: cargo wasi build
       
          Pass --verbose to see the stacktrace.
       
    ✔  nx run vscode-wing:copy (173ms)
    ✔  nx run vscode-wing:build (10s)
    ✔  nx run wingsdk:build (53s)

 ———————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Ran target build for 5 projects and 5 task(s) they depend on (53s)
 
    ✔    5/6 succeeded [2 read from cache]
 
    ✖    1/6 targets failed, including the following:
         - nx run wingc:build-wasm
```

Notice that I don't have wasi in `cargo --list`

Also noticed that we have `wing/.cargo/config.toml` with the following content:
```toml
[env]
CC_wasm32_wasi = "/opt/wasi-sdk/bin/clang"
AR_wasm32_wasi = "/opt/wasi-sdk/bin/ar"
```

While I don't have `opt/wasi-sdk/bin/clang` on my device

**Comment by staycoolcall911** [comment]
Works for me. Make sure you've `git pull`ed and maybe try one more `npm ci` as an offering for the gods of Node?

**Reply by ekeren** [reply]
it didn't help

**Comment by staycoolcall911** [comment]
Where are you running `npm run build` from?

**Reply by ekeren** [reply]
> Where are you running `npm run build` from?

From root project folder

**Comment by ekeren** [comment]
Added information to description

**Comment by staycoolcall911** [comment]
Looks like I have the `cargo wasi` subcommand available and you don't. See `cargo --list | grep wasi`.

**Comment by staycoolcall911** [comment]
Ok, so perhaps `wasi` was installed at some point automatically. Let's see if @3p3r or @MarkMcCulloh can offer some assistance and see if we need to add another installation step to https://github.com/winglang/wing/blob/main/CONTRIBUTING.md

**Reply by ekeren** [reply]
I was able to pass the first phase by running `cargo install wasi` but I am still missing `clang` under `/opt/wasi-sdk/bin/clang`

**Comment by MarkMcCulloh** [comment, marked as answer]
Yeah it looks like the instructions for building the full stack is incomplete now. The readme used to mention that you need to run this script https://github.com/winglang/wing/blob/main/scripts/setup_wasi.sh

`sudo bash scripts/setup_wasi.sh`

**Comment by eladb** [comment]
@MarkMcCulloh any way to automatically install wasi as part of `npm install`?

**Reply by MarkMcCulloh** [reply]
If we improve the idempotency of `scripts/setup_wasi.sh` we can run it as a preinstall hook.
Either that or overhaul our dev setup system.

**Reply by ekeren** [reply]
Is it suppose to run under sudo ?

**Reply by eladb** [reply]
I think it does :-( But why?

(also - before we make this automatic, can we at least add a little check and emit an error with proper instructions on how to run `setup_wasi.sh`?)

**Reply by ekeren** [reply]
It is already added to the CONTRIBUTING.md guide
